src/meta/node/src/server.rs

```diff
@@ -226,123 +235,107 @@ pub async fn rpc_serve(
                 init_system_params,
                 init_session_config,
+                shutdown,
             )
+            .await
         }
     }
 }
 
+/// Bootstraps the follower or leader service based on the election status.
+///
+/// Returns when the `shutdown` token is triggered, or when leader status is lost, or if the leader
+/// service fails to start.
 #[expect(clippy::type_complexity)]
-pub fn rpc_serve_with_store(
+pub async fn rpc_serve_with_store(
     meta_store_impl: MetaStoreImpl,
-    election_client: Option<ElectionClientRef>,
+    election_client: ElectionClientRef,
     address_info: AddressInfo,
     max_cluster_heartbeat_interval: Duration,
     lease_interval_secs: u64,
     opts: MetaOpts,
     init_system_params: SystemParams,
     init_session_config: SessionConfig,
-) -> MetaResult<(JoinHandle<()>, Option<JoinHandle<()>>, WatchSender<()>)> {
-    let (svc_shutdown_tx, svc_shutdown_rx) = watch::channel(());
+    shutdown: CancellationToken,
+) -> MetaResult<()> {
+    // TODO(shutdown): directly use cancellation token
+    let (election_shutdown_tx, election_shutdown_rx) = watch::channel(());
 
-    let leader_lost_handle = if let Some(election_client) = election_client.clone() {
-        let stop_rx = svc_shutdown_tx.subscribe();
+    let election_handle = tokio::spawn({
+        let shutdown = shutdown.clone();
+        let election_client = election_client.clone();
 
-        let handle = tokio::spawn(async move {
+        async move {
             while let Err(e) = election_client
-                .run_once(lease_interval_secs as i64, stop_rx.clone())
+                .run_once(lease_interval_secs as i64, election_shutdown_rx.clone())
                 .await
             {
                 tracing::error!(error = %e.as_report(), "election error happened");
             }
-        });
+            // Leader lost, shutdown the service.
+            shutdown.cancel();
+        }
+    });
 
-        Some(handle)
-    } else {
-        None
-    };
+    if !election_client.is_leader() {
+        let mut is_leader_watcher = election_client.subscribe();
+
+        // If not the leader, spawn a follower.
+        let follower_shutdown = shutdown.child_token();
 
-    let join_handle = tokio::spawn(async move {
-        if let Some(election_client) = election_client.clone() {
-            let mut is_leader_watcher = election_client.subscribe();
-            let mut svc_shutdown_rx_clone = svc_shutdown_rx.clone();
-            let (follower_shutdown_tx, follower_shutdown_rx) = OneChannel::<()>();
+        let follower_handle = tokio::spawn(start_service_as_election_follower(
+            follower_shutdown.clone(),
+            address_info.clone(),
+            Some(election_client.clone()),
+        ));
 
+        // Watch and wait until we become the leader.
+        while !*is_leader_watcher.borrow_and_update() {
             tokio::select! {
-                _ = svc_shutdown_rx_clone.changed() => return,
```

Comment:
Not sure why we always waited for a change before starting the follower service. Is that essentially a synchronization with the election process to confirm the initial status?

I've removed this line and directly check `is_leader()`. May need to add it back if that's true.